### PR TITLE
Decode fragmented requests

### PR DESCRIPTION
**Pull request by cmebarrow**
This is the fix for the request portion of issue #2 Decode fragmented request or response. Decoding fragmented responses will be implemented separately while or after adding support for using the HTTP nukleus as a client.

### CODE REVIEW

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    public void free(int slot)
+    {
+        offsets[slot] = UNUSED;
+        remaining[slot] = 0;
+        acquiredSlots--;
+    }
+
+}
+
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java]: I'd also like to propose a slight iteration on the design, using flyweights.

```
public final class Slab
{
    private final MutableDirectBuffer buffer;
    private final long addressOffset;
    private final int[] offsets;
    private final int bitsPerSlot;

    private final MutableDirectBuffer mutableFW = new UnsafeBuffer(new byte[0]);
    private final DirectBuffer immutableFW = new UnsafeBuffer(new byte[0]);

    public Slab(...)
    {
        this.buffer = new UnsafeBuffer(...);
        this.addressOffset = this.buffer.addressOffset();
        this.offsets = new int[...];
        this.slotCapacity = ...;
        this.bitsPerSlot = ...;
        Arrays.fill(this.offsets, SLOT_AVAILABLE); // -1
        this.availableSlots = this.offsets.length;
        ...
    }

    public int acquire(int key /* streamId */)
    {
        if (availableSlots == 0)
        {
            return SLOT_NOT_AVAILABLE;
        }

        final int mask = offsets.length - 1;
        int slot = Hashing.hash(key, mask);
        while (offsets[slot] != SLOT_AVAILABLE)
        {
            slot = ++slot & mask;
        }
        offsets[slot] = 0;
        availableSlots--;
        return slot;
    }

    public MutableDirectBuffer mutable(int slot)
    {
        final int slotAddressOffset = addressOffset + slot << bitsPerSlot;
        final int offsetWithinSlot = offsets[slot];
        final int length = slotCapacity - offsetWithinSlot;
        return mutableFW.wrap(slotAddressOffset + offsetWithinSlot, length);
    }

    public DirectBuffer immutable(int slot)
    {
        final int slotAddressOffset = addressOffset + slot << bitsPerSlot;
        final int offsetWithinSlot = offsets[slot];
        return immutableFW.wrap(slotAddressOffset, offsetWithinSlot);
    }

   public void release(int slot)
   {
       assert offsets[slot] != SLOT_AVAILABLE;
       offsets[slot] = SLOT_AVAILABLE;
       availableSlots++;
   }
}
```

Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java]: As discussed, let's modify the above to use a `BitSet` to track available slots, and leave mutability outside the `Slab`.

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java [hunk 2]
```diff
@@ -223,17 +230,19 @@ private void processInvalidRequest(
...
+            // TODO: acquire slab for response if targetWindow requires partial write, defer the write till
+            //       we have available (reject) targetWindow
             DirectBuffer payload = new UnsafeBuffer(payloadChars.getBytes(StandardCharsets.UTF_8));
             target.doData(newTargetId, payload, 0, payload.capacity());
 
             this.decoderState = this::decodeHttpBegin;
             this.streamState = this::streamAfterReplyOrReset;
             this.throttleState = this::throttleSkipNextWindow;
-            this.sourceUpdateDeferred = requestBytes - payload.capacity();
+            window += requestBytes;
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: done

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java [hunk 3]
```diff
@@ -304,9 +340,26 @@ private int decodeHttpBegin(
...
+                return limit;
             }
+            else
+            {
+                processCompleteHttpBegin(payload, offset, endOfHeadersAt, endOfHeadersAt - offset);
+                return endOfHeadersAt;
+            }
+        }
 
+        private void processCompleteHttpBegin(
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: done

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java [hunk 4]
```diff
@@ -304,9 +340,26 @@ private int decodeHttpBegin(
...
+            final DirectBuffer payload,
+            final int offset,
+            final int endOfHeadersAt,
+            int windowUpdate)
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: yes, removed

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+ * Methods are provided for getting a buffer and appending partial data in the case of an
+ * incomplete block of data to be decoded. The memory is segmented into slots, the size of
+ * each slot being the maximum allowed size for the header portion of an HTTP request or response.
+ * <b>Each instance of this class is assumed to be used by one and only one thread.</b>
+ */
+class Slab
+{
+    static final int OUT_OF_MEMORY = -2;
+    private static final int UNUSED = -1;
+    private final int slotSize;
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java]: done

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    private final int slotSizeBits;
+    private final MutableDirectBuffer buffer;
+    private final int offsets[]; // starting position for slot data in buffer
+    private final int remaining[]; // length of data in slot
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java]: removed

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    private int acquiredSlots = 0;
+
+    Slab(int minimumNumberOfSlots, int minimumSlotSize)
+    {
+        int slots = findNextPositivePowerOfTwo(minimumNumberOfSlots);
+        slotSize = findNextPositivePowerOfTwo(minimumSlotSize);
+        slotSizeBits = Integer.numberOfTrailingZeros(slotSize);
+        int capacityInBytes = slots << slotSizeBits;
+        buffer = new UnsafeBuffer(new byte[capacityInBytes]);
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java]: done

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        return offsets[slot];
+    }
+
+    public int limit(int slot)
+    {
+        return offsets[slot] + remaining[slot];
+    }
+
+    public int append(int slot, DirectBuffer written, int offset, int limit)
+    {
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java]: method removed

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            slot = ++slot & mask;
+        }
+        offsets[slot] = slot << slotSizeBits;
+        acquiredSlots++;
+        return slot;
+    }
+
+    public void free(int slot)
+    {
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/Slab.java]: done

**Comment by cmebarrow** [comment]
@jfallows Feedback applied. Requires spec changes in https://github.com/reaktivity/nukleus-http.spec/pull/8 plus re-release of http.spec once that PR is merged.

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java [hunk 10]
```diff
@@ -573,7 +566,7 @@ private void throttleSkipNextWindow(
             }
         }
 
-        private void throttleNextWindow(
+        private void throttleDoWindow(
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: `beforeWindowOrReset`

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java [hunk 11]
```diff
@@ -553,17 +549,14 @@ private void handleThrottle(
             throttleState.onMessage(msgTypeId, buffer, index, length);
         }
 
-        private void throttleSkipNextWindow(
+        private void throttleResetOnly(
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: beforeReset

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/Routable.java [hunk 12]
```diff
@@ -166,16 +167,17 @@ private Source newSource(
...
         int maximumHeadersSize = MAXIMUM_HEADERS_SIZE.intValue();
-        int maximumStreamsPendingDecode = MAXIMUM_STREAMS_PENDING_DECODE.intValue(() ->
+        int memoryForDecode = MEMORY_FOR_DECODE.intValue(() ->
         {
-            return (context.maximumStreamsCount() < 1001 ? context.maximumStreamsCount()
-                    : context.maximumStreamsCount() / 10);
+            int maxStreamsWithIncompleteRequest = context.maximumStreamsCount() < 1024 ? context.maximumStreamsCount()
+                    : context.maximumStreamsCount() / 8;
+            return BitUtil.findNextPositivePowerOfTwo(maximumHeadersSize * maxStreamsWithIncompleteRequest / 2);
         });
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/Routable.java]: Added maximumHeadersSize and memoryForDecode methods to Context as discussed

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java (hunk 11 above)
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: done

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java (hunk 10 above)
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: done

File: src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java [hunk 13]
```diff
@@ -593,33 +586,26 @@ private void throttleNextWindow(
             }
         }
 
-        private void processSkipNextWindow(
+        private void processDoWindow(
```
Comment (cmebarrow) [inline src/main/java/org/reaktivity/nukleus/http/internal/routable/stream/SourceInputStreamFactory.java]: done

**Pull request merged**